Experimental okbazel support

aj-michael opened the pull request:
See #232 for context

Documentation for this change is in the commit description of a5a7f9e and in a separate [buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/README.md](https://github.com/uber/okbuck/pull/251/files#diff-4f4f9322475c6d0f4ab2081a290a12c2)

Sorry for the delay on sending this, I had a bit of difficultly getting the cache to work correctly due to some differences between how Buck and Bazel handle resource dependencies.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/okbuck?pullRequest=251) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/okbuck?pullRequest=251) before we can accept your contribution.<br/>

aj-michael commented:
Hmmm, I looked through the Travis CI failure and it doesn't appear to be related to this set of commits.

hzsweers commented:
Thanks! FYI - @kageiit is in Hawaii this week on vacation, so probably won't get reviewed until next week.

aj-michael commented:
Ok, thanks for the heads up 😃

kageiit commented:
Ill take a look by end of this week. Can you check the travis failure?

aj-michael commented:
Sorry for the delayed response, I've been busy this week. I've looked at the Travis failure and it appears unrelated to this change? I'm not sure how to re-trigger Travis to run to check if it is just a flake. As far as I know, my change should not impact any tests as they don't use the OkBazel plugin.

kageiit commented:
I restarted the build. But for future reference, you can close and reopen a PR to have travis rekick your build

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/AndroidBinaryRule.groovy:
```diff
@@ -0,0 +1,42 @@
+
+import com.uber.okbuck.rule.BuckRule
+
+final class AndroidBinaryRule extends BuckRule {
```
Can we name all these classes with a `Bazel` prefix? This is to avoid confusing with the buck classes

aj-michael replied:
Done.

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/AndroidLibraryRule.groovy:
```diff
@@ -0,0 +1,66 @@
+    private final String packageName
+    private final Set<AndroidTarget.ResBundle> resources
+
+    AndroidLibraryRule(
```
does bazel not support annotation processors currently? or is it done implicitly somehow?

aj-michael replied:
Ok, for documentations sake, the way that annotation processors work in Bazel is we create a [`java_plugin`](https://bazel.build/versions/master/docs/be/java.html#java_plugin) rule that depends on a `java_library` that builds that annotation processor or a `java_import` that wraps a prebuilt JAR. The `java_plugin` specifies a `processor_class`. Then we put the `java_plugin` in the deps of an `android_library` or `java_library` that it should run on.

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/AndroidLibraryRuleComposer.groovy:
```diff
@@ -0,0 +1,28 @@
+import com.uber.okbuck.composer.AndroidBuckRuleComposer
+import com.uber.okbuck.core.model.AndroidLibTarget
+
+final class AndroidLibraryRuleComposer extends AndroidBuckRuleComposer {
```
I would suggest extending the `AndroidBuckRuleComposer` to create a `BazelAndroidBuckRuleComposer` which can then override the definitions of `external` etc. to keep the logic cleaner

aj-michael replied:
Ah, I just wrote out a long reply about this and then I think github went down for 2 minutes.

The gist of it was, I've added a commit that introduces BazelRuleComposer. It's a trait as opposed to a subclass of BuckRuleComposer because I need to extend both AndroidBuckRuleComposer and JavaBuckRuleComposer, which each extend BuckRuleComposer.

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/OkBazelGradlePlugin.groovy:
```diff
@@ -0,0 +1,86 @@
+                    PrintStream printer = new PrintStream(subProject.file("BUILD")) {
+                        @Override
+                        public void println(String s) {
+                            // BUILD files are typically space-delimited. Since
```
Great point. I think we should allow customizing the delimiters easily. But this works for now

aj-michael replied:
The Bazel team maintains [buildifier](https://github.com/bazelbuild/buildifier), a tool for formatting BUILD files according to the standard convention. I added this hack so that the BUILD files generated by okbuck adhere to that convention. As an alternative, we could invoke buildifier to clean up the BUILD files at the end. Although, the integration may be messy as buildifier is written in go.

kageiit replied:
I meant customizing delimiters in okbuck, not bazel specifically, but that works as well :)

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/experimental/bazel/README.md:
```diff
@@ -0,0 +1,77 @@
+`WORKSPACE` to contain the `api_level` and `build_tools_version` from your SDK
+that you wish to use.
+
+## Setting up Bazel
```
I would recommend investing in a `bazelw` script that will download and install bazel from source similar to what `buckw` does. It can also run okbuck when files have changed similar to `buckw`. I think this is a very necessary usability feature for adoption.

aj-michael replied:
I will look into how `buckw` works and respond back on this thread. However, I don't think installing bazel from source is a possibility. Bazel can only be built with a pre-existing installation of Bazel. We do this to maintain platform independence (e.g. we don't want to be shipping pre-compiled, platform-specific binaries in the Bazel source tree that would be needed to bootstrap from scratch).

Furthermore, Bazel is very carefully versioned and, since it is pre-1.0, occasionally makes breaking changes. As such, we have some users who intentionally do not update to the newest version immediately because it may require a change to their `BUILD` files. We keep a history of platform-specific installers [here](https://github.com/bazelbuild/bazel/releases) and a detailed changelog for each release [here](https://github.com/bazelbuild/bazel/blob/master/CHANGELOG.md).

I agree that re-running okbuck when files have changed is important for usability.

kageiit replied:
we should atleast have a filed in `bazelw` that will be able to download a particular version of bazel that is prebuilt if it is not buildable from source

aj-michael replied:
Ok, I've added a `bazelWrapper` Gradle task that generates `bazelw`. `bazelw` downloads and executes the installer for bazel 0.4.2 (the most recent stable release at the time of writing). However, it is currently hardcoded to the Linux installer.

aj-michael commented:
I'm still working on addressing all of the comments so far, it's just a busy week for me. I will get to all of them. Are you planning to add more comments than the ones so far?

aj-michael commented:
Ok, I think I have now made it through all of the open comments.

kageiit commented on buildSrc/src/main/resources/com/uber/okbuck/core/util/wrapper/BAZELW_TEMPLATE:
```diff
@@ -0,0 +1,187 @@
+}
+
+# Run tasks before bazel command
+setupBuckRun ( ) {
```
Seems like a lot of common logic is reused across both wrappers. We use a common shell script internally for various wrapper scripts. Will push that change up so this script can get quite small

aj-michael replied:
Yes, the main difference between these scripts is how Buck/Bazel is installed.

kageiit commented:
Can you perform a quick rebase? We made a ton of changes over the past week and want to ensure everything is in working order still

aj-michael commented:
Ok, working on the rebase now. As expected there are some cleanups that I will need to make.

aj-michael commented:
Ok, rebase is complete. I fixed some import paths and confirmed that it still works using `bazelw` on https://github.com/gradle/perf-android-large.

aj-michael commented:
Hey, really sorry about the delay on this. I took a long holiday, and have been catching up on other work in the few days I've been back. I've just finished the rebase and testing to ensure that everything is still good. PTAL :smile:

kageiit commented:
Can you add 

`apply plugin: 'com.uber.okbazel'` to okbuck build.gradle itself.

There are some issues with being able to use both plugins because the bazel plugin is trying to create the same extension as the buck plugin

aj-michael commented:
I asked about sharing the plugin id back in https://github.com/uber/okbuck/issues/232#issuecomment-260689659 , and yes you are correct, this prevents you from using both plugins in the same project. At the time, the reason I did that was because there were several places where the plugin id was used in the codebase and was hardcoded to okbuck.

I can take another look to see what it would take to give it a separate plugin ID.

kageiit commented:
Closing due to inactivity. Feel free to comment anytime if you would like me to re-open this

steeve commented:
@aj-michael did you ever found a way around this? great work though :)
